nslsii/areadetector/xspress3.py

```diff
@@ -0,0 +1,841 @@
+from ophyd import Signal, EpicsSignal, EpicsSignalRO
+
+from ophyd import (
+    Component as Cpt,
+    DynamicDeviceComponent as DynamicDeviceCpt,
+)
+from ophyd.areadetector.filestore_mixins import FileStorePluginBase
+
+from ophyd.areadetector.plugins import HDF5Plugin
+from ophyd.areadetector import ADBase
+from ophyd.areadetector import Xspress3Detector
+from ophyd.device import BlueskyInterface, Staged
+from ophyd.status import DeviceStatus
+
+from databroker.assets.handlers import (
+    Xspress3HDF5Handler,
+    # XS3_XRF_DATA_KEY as XRF_DATA_KEY,
+)
+
+from ..detectors.utils import makedirs
+
+
+logger = logging.getLogger(__name__)
+
+
+class Xspress3Trigger(BlueskyInterface):
+    """ """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._acquire_status = None
+        self._abs_trigger_count = None
+
+    def stage(self):
+        self._abs_trigger_count = 0
+        self.cam.acquire.subscribe(self._acquire_changed)
+        return super().stage()
+
+    def unstage(self):
+        super_unstage_result = super().unstage()
+        self.cam.acquire.clear_sub(self._acquire_changed)
+        self._acquire_status = None
+        return super_unstage_result
+
+    def _acquire_changed(self, value=None, old_value=None, **kwargs):
+        """Respond to changes in the Xspress3Detector.cam.acquire PV.
+
+        The important behavior of this method is to mark self._acquire_status
+        as `finished` when Xspress3Detector.cam.acquire changes from 1 to 0
+        (high to low).
+
+        Parameters
+        ----------
+        value: int
+            new value of the Xspress3Detector.cam.acquire PV
+        old_value: int
+            previous value of the Xspress3Detector.cam.acquire PV
+        kwargs: dict
+            unused
+
+        Returns
+        -------
+        No return value
+        """
+
+        if self._acquire_status is None:
+            return
+        if (old_value == 1) and (value == 0):
+            # Negative-going edge means an acquisition just finished.
+            self._acquire_status.set_finished()
+            self._acquire_status = None
+
+    def new_acquire_status(self):
+        """
+        Create, remember, and return a Status object that will be marked
+        as `finished` when acquisition is done (see _acquire_changed). The
+        intention is that this Status will be used by another object,
+        for example a RunEngine.
+
+        Returns
+        -------
+        DeviceStatus
+        """
+
+        self._acquire_status = DeviceStatus(self)
+        return self._acquire_status
+
+    def trigger(self):
+        if self._staged != Staged.yes:
+            raise RuntimeError(
+                "tried to trigger Xspress3 with prefix {self.prefix} but it is not staged"
+            )
+
+        acquire_status = self.new_acquire_status()
+        self.cam.acquire.put(1, wait=False)
```

Comment:
I think I used `put` because it was in the original code. `set` does a lot of work that doesn't seem necessary here since there is already a Status object watching the acquire PV.

Reply:
After consulting the oracle I believe `put` is the correct method here since it is part of the ophyd EPICS interface, while `set` is intended for a higher level client such as bluesky.